Tracking Issue for scoped threads

Feature gate: `#![feature(scoped_threads)]`

This is a tracking issue for scoped threads.

RFC: https://rust-lang.github.io/rfcs/3151-scoped-threads.html

**Note:** Right now, some usage patterns might need a `#![feature(nll)]` to avoid lifetime errors. (This is probably stable soon.)

### Example usage

```rust
let local_var = vec![1, 2, 3];

thread::scope(|s| {
    s.spawn(|| println!("borrowed from thread #1: {:?}", local_var));
    s.spawn(|| println!("borrowed from thread #2: {:?}", local_var));
    println!("borrowed from the main thread: {:?}", local_var);
});
```

### Public API

Documentation: https://doc.rust-lang.org/nightly/std/thread/fn.scope.html

```rust
// std::thread

pub fn scope<'env, F, T>(f: F) -> T
where
    F: for<'scope> FnOnce(&'scope Scope<'scope, 'env>) -> T;

pub struct Scope<'scope, 'env: 'scope> { ... }

impl<'scope, 'env> Scope<'scope, 'env> {
    pub fn spawn<F, T>(&'scope self, f: F) -> ScopedJoinHandle<'scope, T>
    where
        F: FnOnce() -> T + Send + 'scope,
        T: Send + 'scope;
}

impl Builder {
    pub fn spawn_scoped<'scope, 'env, F, T>(
        self,
        scope: &'scope Scope<'env>,
        f: F,
    ) -> io::Result<ScopedJoinHandle<'scope, T>>
    where
        F: FnOnce() -> T + Send + 'scope,
        T: Send + 'scope;
}

pub struct ScopedJoinHandle<'scope, T> { ... }

impl<'scope, T> ScopedJoinHandle<'scope, T> {
    pub fn join(self) -> Result<T>;
    pub fn thread(&self) -> &Thread;
    pub fn is_finshed(&self) -> bool;
}
```

### Steps / History

- [x] RFC attempt 1: https://github.com/rust-lang/rfcs/pull/2647
- [x] RFC attempt 2: https://github.com/rust-lang/rfcs/pull/3151
- [x] Implementation: https://github.com/rust-lang/rust/pull/92555
- [x] Change signatures a bit to remove the argument to the spawn closures: https://github.com/rust-lang/rust/pull/94559
- [x] Fix soundness issue in implementation: https://github.com/rust-lang/rust/pull/94644
- [x] Document lifetimes: https://github.com/rust-lang/rust/pull/94763
- [x] Final comment period (FCP): https://github.com/rust-lang/rust/issues/93203#issuecomment-1075284811
- [x] Stabilization PR: https://github.com/rust-lang/rust/pull/97992

### Unresolved Questions

- [x] Can we omit the `&Scope` argument to the functions given to `.spawn()`? That is, `scope.spawn(|| ..)` rather than `scope.spawn(|_| ..)`.
  - It's already possible by forcing the user to use `move || ..` instead, but that's not great. Maybe the language could be subtly changed to capture references or certain `Copy` types by value rather than by reference(-to-reference).
  - See also [this comment](https://github.com/rust-lang/rfcs/pull/3151#issuecomment-883251161) and the collapsed section in [the comment below](https://github.com/rust-lang/rust/issues/93203#issuecomment-1019293467).
  - Mostly answered in [this comment](https://github.com/rust-lang/rust/issues/93203#issuecomment-1040720703).
  - Working idea in [this comment](https://github.com/rust-lang/rust/issues/93203#issuecomment-1040798286).
  - Implementation in https://github.com/rust-lang/rust/pull/94559
- [x] How to document the `'env` and `'scope` lifetimes clearly without scaring people away.
  - https://github.com/rust-lang/rust/pull/94763

Awesome API, thanks again @m-ou-se for taking care of driving this 👍

<details><summary>To expand on my captures suggestion / "thingy"</summary>

I think it would _generally_ make sense to be able to mark (`Copy`) type definitions with a `#[favor_capture_by_move]` annotation on them or something, so that, in a non-generic context, if one has `let x = instance_of_that type; || uses(x) || uses(&x)`, `x` is nonetheless captured by move.
It sounds like a kind of straight-forward thing to implement? I'd be willing to try to tackle the implementation 🙂.
With it, we'd then replace `&'scope Scope<'env>` with `ScopeRef<'scope, 'env>` (probably still named `Scope`, but I'm using the `Ref` suffix here to clarify the comparison with the current implementation). The reason for that would be to have that first `'scope` parameter be non-covariant (≠ the (co)variance of `'scope` in `&'scope Scope<'env>`), so that the requirement for the scoped joined thread closure would become `F : 'scope` rather than `F : 'env`, so that `F` would be allowed to capture the `ScopeRef`.

With all that, the API would become:

```rust
thread::scope(|scope| {
    …
    scope.spawn(|| {
        …
        scope.spawn(|| { … });
    });
    …
});
```

which is not only neater (there is one _scoped_ / scope-defining API, at the outermost closure layer, and then good old `.spawn(|| …)` functions), but would also avoid the rather hard to grasp error messages that somebody gets when they update their scoped threads usage with a new nested thread.

Indeed, consider somebody updating some code and doing:

```diff
fn main ()
{
    ::crossbeam::thread::scope(|scope| {
        scope.spawn(|_| {
            // …
        });
        scope.spawn(|_| {
+           scope.spawn(|_| { /* … */ })
            // …
        });
    })
    .expect("Some thread panicked");
}
```

They'd get, as of the current `nightly`, the following error message:

```rust
error[E0521]: borrowed data escapes outside of closure
  --> src/main.rs:7:9
   |
3  |       ::crossbeam::thread::scope(|scope| {
   |                                   -----
   |                                   |
   |                                   `scope` declared here, outside of the closure body
   |                                   `scope` is a reference that is only valid in the closure body
...
7  | /         scope.spawn(|_| {
8  | |             scope.spawn(|_| { /* … */ }) // <- ADDED!
9  | |             // …
10 | |         });
   | |__________^ `scope` escapes the closure body here
   |
   = note: requirement occurs because of the type Scope<'_>, which makes the generic argument '_ invariant
   = note: the struct Scope<'env> is invariant over the parameter 'env
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
```

  - [Demo](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=dba094d4415f59d94a81908297c1795a)

Now, I personally do love the quality of this error message, but I think we have to admit it can still be quite too "technical / advanced" for some Rust programmers out there (Alan yadda yadda): they shouldn't have to know about variance to know how to use a scoped API. Granted, another option would be to invest some effort in improving the error message when using a scoped API / hard-coding this very situation, but I'd rather see that effort invested on `#[favor_capture_by_move]`, for instance.

___

These have been my two cents on the topic, thanks for reading 🙏, and won't insist anymore on this subject, I promise 😄

</details>

--- a/library/std/src/thread/tests.rs
+++ b/library/std/src/thread/tests.rs
@@ -316,3 +316,16 @@ fn test_scoped_threads_drop_result_before_join() {
     });
     assert!(actually_finished.load(Ordering::Relaxed));
 }
+
+#[test]
+fn test_scoped_threads_nll() {
+    // this is mostly a *compilation test* for this exact function:
+    fn foo(x: &u8) {
+        thread::scope(|s| {
+            s.spawn(|| drop(x));
+        });
+    }
+    // let's also run it for good measure
+    let x = 42_u8;
+    foo(&x);
+}

